### ISSUE TITLE
EventRestarters do not retain observer configuration (and other footguns)

### ISSUE THREAD

**Issue by VTCAKAVSMoACE** [AFLplusplus/LibAFL#925]
There seems to be an undocumented footgun w.r.t. Observers where they can retain configuration values (e.g., `initial` value for `MapObserver`s) which isn't restored on restart by `EventRestarter`s, for the "obvious" reason that it isn't present in state metadata and won't be retained.

Some of the existing implementations seem to assume that this data is retained, or don't consider that it won't. Additionally, information on what is *not* retained by EventRestarter isn't present.

Perhaps it would be worthwhile to do the following:
 - Document information that should be retained by `EventRestarter`s
 - Update `Observer` traits to ensure that configuration data is retained in state metadata
 - An example fuzzer (or several) which shows erroneous use of various features ("what not to do")

**Comment by domenukk** [comment]
I don't quite understand what the issue is? What does your configuration look like that you lose on restart?

**Comment by VTCAKAVSMoACE** [comment]
If the content of an observer changes during fuzzing, then a restart happens, the observer content is discarded (which is fine when all it contains is the current run's data). This is an issue if the observer is configured during fuzzing. This is already the case for most MapObservers which have the `initial` field, which can be changed during runtime but without this change being propagated through restarts.

**Comment by domenukk** [comment]
Ah I see, so they should really be writing that to the state..
Do they have that option?

**Comment by VTCAKAVSMoACE** [comment]
Currently no, since this configuration is provided via mutable reference. The API would need to change, and this would require a lookup on *every* execution (which would be not great). Maybe observers (and other types) need to be notified when the state is restored, and cache metadata?

**Comment by domenukk** [comment]
Yes it might be good to have a "Metadata Updated" callback or semething?

**Comment by VTCAKAVSMoACE** [comment]
Or the opposite: state restored callback. The observers will need to update metadata themselves regardless (on pre-exec?).

**Comment by domenukk** [comment]
A metadata changed callback would drastically improve the performance of tuneable mutators and schedulers
https://github.com/AFLplusplus/LibAFL/blob/e5aaf85d3cc0e0000b76022446ef9a36f0803476/libafl/src/schedulers/tuneable.rs

**Comment by VTCAKAVSMoACE** [comment]
How would this callback be implemented? Would we register a map with typeids and a list of listeners?

**Comment by VTCAKAVSMoACE** [comment]
Also, I see a large PR in libafl's near future :sweat_smile: There's a lot of types that will need to be updated.

Oh! And we could simply walk all of the metadata callbacks immediately after a state restore. I think this was what I was misunderstanding from your solution earlier.

**Comment by domenukk** [comment]
I don't have a good answer, I guess we need to walk all elements of the fuzzer once, if possible?

**Comment by VTCAKAVSMoACE** [comment]
What about storing metadata in Rc<RefMut<_>> instead of Box<_>? Then we could keep (updated) references to metadata present throughout the fuzzer without needing to re-query it.

**Comment by domenukk** [comment]
Could be an issue for Serde

**Comment by domenukk** [comment]
closing this after inactivity (reopen if still relevant)
